pkg/utils/retry.go

```diff
@@ -0,0 +1,55 @@
+package utils
+
+import (
+	"math/rand"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type retryError struct {
+	error
+}
+
+// Retry attempts to execute the given callback function a certain number of times,
+// with a delay between each attempt using a simple exponential backoff algorithm
+// that uses a slight jitter to ensure that retries aren't clustered.
+//
+// If the callback function returns an error, Retry will sleep for a certain duration
+// before attempting to call the callback function again.
+//
+// Retry will repeat this process until the callback function succeeds or until the
+// maximum number of attempts is reached.
+//
+// If the maximum number of attempts is reached and the callback function still fails,
+// it returns the last error returned by the callback function.
+func Retry(attempts int, sleep time.Duration, callbackFunc func() error) error {
```

Comment:
Please reconsider if this can be done using the hashicorp vault client.

Reply:
@janboll, the client retries failures on the transport layer (HTTP or HTTPS), but empty authentication information is something one has to take care of, unfortunately. This retry here aims to make obtaining credentials more robust.

If you don't mind, I will keep it and switch it to the new API, as mentioned in the above comment.

Thank you!